src/api/server.rs

```diff
@@ -0,0 +1,123 @@
+use super::{handler, schema::build_schema};
+
+use async_graphql::{
+    http::{playground_source, GraphQLPlaygroundConfig},
+    QueryBuilder,
+};
+use async_graphql_warp::{graphql_subscription, GQLResponse};
+use std::convert::Infallible;
+use std::net::SocketAddr;
+use tokio::sync::oneshot;
+use tokio::sync::oneshot::{Receiver, Sender};
+use warp::filters::BoxedFilter;
+use warp::{http::Response, Filter, Reply};
+
+pub struct Server {
+    /// Address for the API server to bind on
+    address: SocketAddr,
+
+    /// Transmission channel to trigger closure of a running API server
+    trigger_cancel: Sender<()>,
+
+    /// Receiver signal to cancel running API server
+    cancel_signal: Option<Receiver<()>>,
+}
+
+impl Server {
+    /// Returns a new API Server
+    pub fn new(address: SocketAddr) -> Server {
+        let (trigger_cancel, cancel_signal) = oneshot::channel::<()>();
+
+        Server {
+            address,
+            trigger_cancel,
+            cancel_signal: Some(cancel_signal),
+        }
+    }
+
+    /// String representation of the bound IP address
+    pub fn ip(&self) -> String {
```

Comment:
It's more idiomatic to return a reference to a string (`&str`). This removes the forced allocation, but still allows the callee to decide if it needs an owned string by calling `to_string()` itself.

Reply:
Ah, never mind. There is no way to get a string representation of `IpAddr` without allocating.

Reply:
This could be a reason to have this function return the `IpAddr` type instead, and have the callee call `to_string` if needed. Similar to our discussion about `port()`.

But having said that, since this is only used in one location, I don't mind too much what choice we make.

Reply:
Updated locally to return the underlying types, and require the callee to convert as required:

Usage is now:

```rust
info!(
    message = format!("{} API server", message).as_str(),
    ip = server.ip().to_string().as_str(),
    port = server.port().to_string().as_str()
);
```

The `.to_string().as_str()` hope is there to satisfy the `tracing::Value` trait bound, which isn't implement for `String`.

Reply:
Looks good to me 👍

I suspect `port = &server.port().to_string()` should work as wel, as `String` derefences to `&str`.

See also: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=741616df7f3cb47af85b219779d654fa

Reply:
This is to satisfy `tracing::Value`:

```rust
error[E0277]: the trait bound `std::string::String: tracing::Value` is not satisfied
   --> src/topology/mod.rs:493:9
    |
493 | /         info!(
494 | |             message = format!("{} API server", message).as_str(),
495 | |             ip = server.ip().to_string(),
496 | |             port = server.port().to_string().as_str()
497 | |         );
    | |__________^ the trait `tracing::Value` is not implemented for `std::string::String`

Reply:
> This is to satisfy `tracing::Value`

I understand, what I suggested was applying [deref coercion](https://doc.rust-lang.org/nomicon/coercions.html):

> Expression &x of type &T to &*x of type &U if T derefs to U (i.e. T: Deref<Target=U>)

since `impl Deref<Target=str> for String` [exists](https://doc.rust-lang.org/1.45.2/std/string/struct.String.html#impl-Deref), that gets rid of the extra `as_str()`. Anyway, it's a stylistic difference, so there's no need to change anything here, just wanted to point this out for future use 👍.